{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f2071634-bd0c-4f7c-8588-cf3f670f2b8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c44ef4a3-84cd-4815-b784-8ee377157199",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import ast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ea81eafb-d88f-4ea7-ac37-a351c23c91a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "BACDIVE_DIR = Path.cwd().parent / \"kg_microbe/transform_utils/bacdive\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "78ff9a75-bea9-4ac8-9b5d-93d812bf5a48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bacdive_id</th>\n",
       "      <th>dsm_number</th>\n",
       "      <th>culture_collection_number</th>\n",
       "      <th>ncbitaxon_id</th>\n",
       "      <th>ncbitaxon_description</th>\n",
       "      <th>keywords</th>\n",
       "      <th>medium_id</th>\n",
       "      <th>medium_label</th>\n",
       "      <th>medium_url</th>\n",
       "      <th>mediadive_medium_url</th>\n",
       "      <th>isolation</th>\n",
       "      <th>isolation source categories</th>\n",
       "      <th>Morphology_multimedia</th>\n",
       "      <th>Morphology_multicellular morphology</th>\n",
       "      <th>Morphology_colony morphology</th>\n",
       "      <th>Morphology_cell morphology</th>\n",
       "      <th>Morphology_pigmentation</th>\n",
       "      <th>risk assessment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bacdive:168835</td>\n",
       "      <td>28028.0</td>\n",
       "      <td>['LMG 27950', ' DSM 28028']</td>\n",
       "      <td>NCBITaxon:561122</td>\n",
       "      <td>Paraburkholderia peleae PP52-1 is a bacterium ...</td>\n",
       "      <td>['16S sequence', 'Bacteria']</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'@ref': 20215, 'biosafety level': '1', 'biosa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bacdive:168565</td>\n",
       "      <td>25187.0</td>\n",
       "      <td>['DSM 25187', ' KCCM 90101']</td>\n",
       "      <td>NCBITaxon:1075417</td>\n",
       "      <td>Catalinimonas alkaloidigena CNU-914 is a bacte...</td>\n",
       "      <td>['16S sequence', 'Bacteria']</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'@ref': 20215, 'biosafety level': '1', 'biosa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bacdive:167805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['NBRC 114727', ' KACC 22074']</td>\n",
       "      <td>NCBITaxon:2801341</td>\n",
       "      <td>Noviherbaspirillum pedocola DKR-6 is a bacteri...</td>\n",
       "      <td>['16S sequence', 'Bacteria']</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bacdive:165157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['JCM 5373']</td>\n",
       "      <td>NCBITaxon:296</td>\n",
       "      <td>Pseudomonas fragi JCM 5373 is a mesophilic bac...</td>\n",
       "      <td>['Bacteria', '16S sequence', 'mesophilic']</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'@ref': 67770, 'sample type': 'Pork'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bacdive:164742</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['JCM 34174', ' GDMCC 1.1677', ' KACC 21469']</td>\n",
       "      <td>NCBITaxon:2751478</td>\n",
       "      <td>Janthinobacterium rivuli JCM 34174 is a psychr...</td>\n",
       "      <td>['psychrophilic', 'Bacteria']</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       bacdive_id  dsm_number                      culture_collection_number  \\\n",
       "0  bacdive:168835     28028.0                    ['LMG 27950', ' DSM 28028']   \n",
       "1  bacdive:168565     25187.0                   ['DSM 25187', ' KCCM 90101']   \n",
       "2  bacdive:167805         NaN                 ['NBRC 114727', ' KACC 22074']   \n",
       "3  bacdive:165157         NaN                                   ['JCM 5373']   \n",
       "4  bacdive:164742         NaN  ['JCM 34174', ' GDMCC 1.1677', ' KACC 21469']   \n",
       "\n",
       "        ncbitaxon_id                              ncbitaxon_description  \\\n",
       "0   NCBITaxon:561122  Paraburkholderia peleae PP52-1 is a bacterium ...   \n",
       "1  NCBITaxon:1075417  Catalinimonas alkaloidigena CNU-914 is a bacte...   \n",
       "2  NCBITaxon:2801341  Noviherbaspirillum pedocola DKR-6 is a bacteri...   \n",
       "3      NCBITaxon:296  Pseudomonas fragi JCM 5373 is a mesophilic bac...   \n",
       "4  NCBITaxon:2751478  Janthinobacterium rivuli JCM 34174 is a psychr...   \n",
       "\n",
       "                                     keywords medium_id medium_label  \\\n",
       "0                ['16S sequence', 'Bacteria']       NaN          NaN   \n",
       "1                ['16S sequence', 'Bacteria']       NaN          NaN   \n",
       "2                ['16S sequence', 'Bacteria']       NaN          NaN   \n",
       "3  ['Bacteria', '16S sequence', 'mesophilic']       NaN          NaN   \n",
       "4               ['psychrophilic', 'Bacteria']       NaN          NaN   \n",
       "\n",
       "  medium_url mediadive_medium_url                               isolation  \\\n",
       "0        NaN                  NaN                                     NaN   \n",
       "1        NaN                  NaN                                     NaN   \n",
       "2        NaN                  NaN                                     NaN   \n",
       "3        NaN                  NaN  {'@ref': 67770, 'sample type': 'Pork'}   \n",
       "4        NaN                  NaN                                     NaN   \n",
       "\n",
       "  isolation source categories Morphology_multimedia  \\\n",
       "0                         NaN                   NaN   \n",
       "1                         NaN                   NaN   \n",
       "2                         NaN                   NaN   \n",
       "3                         NaN                   NaN   \n",
       "4                         NaN                   NaN   \n",
       "\n",
       "  Morphology_multicellular morphology Morphology_colony morphology  \\\n",
       "0                                 NaN                          NaN   \n",
       "1                                 NaN                          NaN   \n",
       "2                                 NaN                          NaN   \n",
       "3                                 NaN                          NaN   \n",
       "4                                 NaN                          NaN   \n",
       "\n",
       "  Morphology_cell morphology Morphology_pigmentation  \\\n",
       "0                        NaN                     NaN   \n",
       "1                        NaN                     NaN   \n",
       "2                        NaN                     NaN   \n",
       "3                        NaN                     NaN   \n",
       "4                        NaN                     NaN   \n",
       "\n",
       "                                     risk assessment  \n",
       "0  {'@ref': 20215, 'biosafety level': '1', 'biosa...  \n",
       "1  {'@ref': 20215, 'biosafety level': '1', 'biosa...  \n",
       "2                                                NaN  \n",
       "3                                                NaN  \n",
       "4                                                NaN  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(BACDIVE_DIR/\"tmp/bacdive.tsv\", sep=\"\\t\", low_memory=False)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f9273d5f-d5f5-4b7a-bcbc-db8fd64d981f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['16S sequence', 'Bacteria', 'mesophilic', 'psychrophilic', 'genome sequence', 'thermophilic', 'Gram-negative', 'aerobe', 'anaerobe', 'obligate aerobe', 'rod-shaped', 'colony-forming', 'Gram-positive', 'motile', 'microaerophile', 'human pathogen', 'Archaea', 'sphere-shaped', 'spore-forming', 'facultative anaerobe', 'moderately halophilic', 'antibiotic compound production', 'animal pathogen', 'coccus-shaped', 'pigmented', 'Gram-variable', 'obligate anaerobe', 'spiral-shaped', 'plant pathogen', 'ovoid-shaped', 'filament-shaped', 'antibiotic resistance', 'chemoorganotroph', 'halophilic', 'non-halophilic', 'chemoheterotroph', 'alcohol production', 'other', 'chemoorganoheterotroph', 'heterotroph', 'oval-shaped', 'curved-shaped', 'hyperthermophilic', 'pleomorphic-shaped', 'chemoautolithotroph', 'toxin production', 'facultative aerobe', 'methylotroph', 'amino acid production', 'halotolerant', 'helical-shaped', 'mixotroph', 'chemolithoautotroph', 'aerotolerant', 'methanotroph', 'lithoheterotroph', 'extremely halophilic', 'organoheterotroph', 'vibrio-shaped', 'lactate production', 'polysaccharide production', 'chemolithotroph', 'microaerotolerant', 'crescent-shaped', 'autotroph', 'slightly halophilic', 'methane production', 'phototroph', 'photoheterotroph', 'haloalkaliphilic', 'star-shaped', 'ellipsoidal', 'lithoautotroph', 'diazotroph', 'chemolithoheterotroph', 'chemotroph', 'organotroph', 'spore-shaped', 'diplococcus-shaped', 'photolithotroph', 'organotroph|photoautotroph', 'dumbbell-shaped', 'photoorganoheterotroph', 'lithotroph', 'oligotroph', 'chemoautotroph', 'ring-shaped']\n"
     ]
    }
   ],
   "source": [
    "# Step 1: Convert stringified lists to actual lists\n",
    "df['keywords'] = df['keywords'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) and x.startswith('[') else x)\n",
    "\n",
    "# Ensure all elements are lists (for any remaining single string entries)\n",
    "df['keywords'] = df['keywords'].apply(lambda x: x if isinstance(x, list) else [x])\n",
    "\n",
    "# Step 2: Explode the 'keywords' column\n",
    "exploded_df = df.explode('keywords')\n",
    "\n",
    "# Step 3: Get unique values\n",
    "unique_keywords = exploded_df['keywords'].unique()\n",
    "\n",
    "# Step 4: Convert to list\n",
    "unique_keywords_list = unique_keywords.tolist()\n",
    "\n",
    "# Display the list of unique keywords\n",
    "print(unique_keywords_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "db203e3b-8aec-490a-af84-028cb868bf46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bacdive_id</th>\n",
       "      <th>observation</th>\n",
       "      <th>enzymes</th>\n",
       "      <th>metabolite utilization</th>\n",
       "      <th>metabolite production</th>\n",
       "      <th>metabolite tests</th>\n",
       "      <th>API_X</th>\n",
       "      <th>oxygen tolerance</th>\n",
       "      <th>spore formation</th>\n",
       "      <th>halophily</th>\n",
       "      <th>antibiotic resistance</th>\n",
       "      <th>murein</th>\n",
       "      <th>compound production</th>\n",
       "      <th>fatty acid profile</th>\n",
       "      <th>tolerance</th>\n",
       "      <th>antibiogram</th>\n",
       "      <th>nutrition type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bacdive:164333</td>\n",
       "      <td>{'@ref': 67770, 'observation': 'quinones: Q-8'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bacdive:163937</td>\n",
       "      <td>{'@ref': 67770, 'observation': 'quinones: MK-7'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bacdive:163735</td>\n",
       "      <td>{'@ref': 67770, 'observation': 'quinones: MK-8...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bacdive:162698</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'@ref': 67770, 'value': 'dextransucrase', 'ec...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bacdive:159798</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'@ref': 66741, 'Chebi-ID': 16634, 'metabolit...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       bacdive_id                                        observation  \\\n",
       "0  bacdive:164333    {'@ref': 67770, 'observation': 'quinones: Q-8'}   \n",
       "1  bacdive:163937   {'@ref': 67770, 'observation': 'quinones: MK-7'}   \n",
       "2  bacdive:163735  {'@ref': 67770, 'observation': 'quinones: MK-8...   \n",
       "3  bacdive:162698                                                NaN   \n",
       "4  bacdive:159798                                                NaN   \n",
       "\n",
       "                                             enzymes  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3  {'@ref': 67770, 'value': 'dextransucrase', 'ec...   \n",
       "4                                                NaN   \n",
       "\n",
       "                              metabolite utilization metabolite production  \\\n",
       "0                                                NaN                   NaN   \n",
       "1                                                NaN                   NaN   \n",
       "2                                                NaN                   NaN   \n",
       "3                                                NaN                   NaN   \n",
       "4  [{'@ref': 66741, 'Chebi-ID': 16634, 'metabolit...                   NaN   \n",
       "\n",
       "  metabolite tests API_X oxygen tolerance spore formation halophily  \\\n",
       "0              NaN   NaN              NaN             NaN       NaN   \n",
       "1              NaN   NaN              NaN             NaN       NaN   \n",
       "2              NaN   NaN              NaN             NaN       NaN   \n",
       "3              NaN   NaN              NaN             NaN       NaN   \n",
       "4              NaN   NaN              NaN             NaN       NaN   \n",
       "\n",
       "  antibiotic resistance murein compound production fatty acid profile  \\\n",
       "0                   NaN    NaN                 NaN                NaN   \n",
       "1                   NaN    NaN                 NaN                NaN   \n",
       "2                   NaN    NaN                 NaN                NaN   \n",
       "3                   NaN    NaN                 NaN                NaN   \n",
       "4                   NaN    NaN                 NaN                NaN   \n",
       "\n",
       "  tolerance antibiogram nutrition type  \n",
       "0       NaN         NaN            NaN  \n",
       "1       NaN         NaN            NaN  \n",
       "2       NaN         NaN            NaN  \n",
       "3       NaN         NaN            NaN  \n",
       "4       NaN         NaN            NaN  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.read_csv(BACDIVE_DIR/\"tmp/bacdive_physiology_metabolism.tsv\", sep=\"\\t\", low_memory=False)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "cc9f8b61-5ba6-44d3-80a1-76a03f726ed2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bacdive_id</th>\n",
       "      <th>observation</th>\n",
       "      <th>enzymes</th>\n",
       "      <th>metabolite utilization</th>\n",
       "      <th>metabolite production</th>\n",
       "      <th>metabolite tests</th>\n",
       "      <th>API_X</th>\n",
       "      <th>oxygen tolerance</th>\n",
       "      <th>spore formation</th>\n",
       "      <th>halophily</th>\n",
       "      <th>antibiotic resistance</th>\n",
       "      <th>murein</th>\n",
       "      <th>compound production</th>\n",
       "      <th>fatty acid profile</th>\n",
       "      <th>tolerance</th>\n",
       "      <th>antibiogram</th>\n",
       "      <th>nutrition type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>bacdive:156647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'@ref': 63037, 'oxygen tolerance': 'aerobe'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>bacdive:156247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'@ref': 68380, 'value': 'serine arylamidase'...</td>\n",
       "      <td>[{'@ref': 68380, 'Chebi-ID': 27897, 'metabolit...</td>\n",
       "      <td>{'@ref': 68380, 'Chebi-ID': 35581, 'metabolite...</td>\n",
       "      <td>{'@ref': 68380, 'Chebi-ID': 35581, 'metabolite...</td>\n",
       "      <td>{'API zym': {'@ref': 62467, 'Control': '-', 'A...</td>\n",
       "      <td>{'@ref': 62467, 'oxygen tolerance': 'anaerobe'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>bacdive:155800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'@ref': 68374, 'value': 'L-aspartate arylami...</td>\n",
       "      <td>[{'@ref': 68374, 'Chebi-ID': 30911, 'metabolit...</td>\n",
       "      <td>{'@ref': 68374, 'Chebi-ID': 35581, 'metabolite...</td>\n",
       "      <td>{'@ref': 68374, 'Chebi-ID': 35581, 'metabolite...</td>\n",
       "      <td>{'API ID32E': {'@ref': 61922, 'ODC': '+', 'ADH...</td>\n",
       "      <td>{'@ref': 61922, 'oxygen tolerance': 'aerobe'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>bacdive:155698</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'@ref': 61807, 'oxygen tolerance': 'anaerobe'}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>bacdive:152517</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'@ref': 68381, 'value': 'urease', 'activity'...</td>\n",
       "      <td>[{'@ref': 68380, 'Chebi-ID': 27897, 'metabolit...</td>\n",
       "      <td>[{'@ref': 68380, 'Chebi-ID': 35581, 'metabolit...</td>\n",
       "      <td>[{'@ref': 68380, 'Chebi-ID': 35581, 'metabolit...</td>\n",
       "      <td>{'API rID32A': {'@ref': 57438, 'URE': '-', 'AD...</td>\n",
       "      <td>[{'@ref': 57438, 'oxygen tolerance': 'anaerobe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        bacdive_id observation  \\\n",
       "6   bacdive:156647         NaN   \n",
       "7   bacdive:156247         NaN   \n",
       "8   bacdive:155800         NaN   \n",
       "9   bacdive:155698         NaN   \n",
       "10  bacdive:152517         NaN   \n",
       "\n",
       "                                              enzymes  \\\n",
       "6                                                 NaN   \n",
       "7   [{'@ref': 68380, 'value': 'serine arylamidase'...   \n",
       "8   [{'@ref': 68374, 'value': 'L-aspartate arylami...   \n",
       "9                                                 NaN   \n",
       "10  [{'@ref': 68381, 'value': 'urease', 'activity'...   \n",
       "\n",
       "                               metabolite utilization  \\\n",
       "6                                                 NaN   \n",
       "7   [{'@ref': 68380, 'Chebi-ID': 27897, 'metabolit...   \n",
       "8   [{'@ref': 68374, 'Chebi-ID': 30911, 'metabolit...   \n",
       "9                                                 NaN   \n",
       "10  [{'@ref': 68380, 'Chebi-ID': 27897, 'metabolit...   \n",
       "\n",
       "                                metabolite production  \\\n",
       "6                                                 NaN   \n",
       "7   {'@ref': 68380, 'Chebi-ID': 35581, 'metabolite...   \n",
       "8   {'@ref': 68374, 'Chebi-ID': 35581, 'metabolite...   \n",
       "9                                                 NaN   \n",
       "10  [{'@ref': 68380, 'Chebi-ID': 35581, 'metabolit...   \n",
       "\n",
       "                                     metabolite tests  \\\n",
       "6                                                 NaN   \n",
       "7   {'@ref': 68380, 'Chebi-ID': 35581, 'metabolite...   \n",
       "8   {'@ref': 68374, 'Chebi-ID': 35581, 'metabolite...   \n",
       "9                                                 NaN   \n",
       "10  [{'@ref': 68380, 'Chebi-ID': 35581, 'metabolit...   \n",
       "\n",
       "                                                API_X  \\\n",
       "6                                                 NaN   \n",
       "7   {'API zym': {'@ref': 62467, 'Control': '-', 'A...   \n",
       "8   {'API ID32E': {'@ref': 61922, 'ODC': '+', 'ADH...   \n",
       "9                                                 NaN   \n",
       "10  {'API rID32A': {'@ref': 57438, 'URE': '-', 'AD...   \n",
       "\n",
       "                                     oxygen tolerance spore formation  \\\n",
       "6       {'@ref': 63037, 'oxygen tolerance': 'aerobe'}             NaN   \n",
       "7     {'@ref': 62467, 'oxygen tolerance': 'anaerobe'}             NaN   \n",
       "8       {'@ref': 61922, 'oxygen tolerance': 'aerobe'}             NaN   \n",
       "9     {'@ref': 61807, 'oxygen tolerance': 'anaerobe'}             NaN   \n",
       "10  [{'@ref': 57438, 'oxygen tolerance': 'anaerobe...             NaN   \n",
       "\n",
       "   halophily antibiotic resistance murein compound production  \\\n",
       "6        NaN                   NaN    NaN                 NaN   \n",
       "7        NaN                   NaN    NaN                 NaN   \n",
       "8        NaN                   NaN    NaN                 NaN   \n",
       "9        NaN                   NaN    NaN                 NaN   \n",
       "10       NaN                   NaN    NaN                 NaN   \n",
       "\n",
       "   fatty acid profile tolerance antibiogram nutrition type  \n",
       "6                 NaN       NaN         NaN            NaN  \n",
       "7                 NaN       NaN         NaN            NaN  \n",
       "8                 NaN       NaN         NaN            NaN  \n",
       "9                 NaN       NaN         NaN            NaN  \n",
       "10                NaN       NaN         NaN            NaN  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "o2_tolerance_filtered_df = df2[df2['oxygen tolerance'].notna()]\n",
    "o2_tolerance_filtered_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "7846c837-1672-444c-a327-c485527bf27c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to extract 'oxygen tolerance' from a dict or list of dicts\n",
    "def extract_oxygen_tolerance(data):\n",
    "    if isinstance(data, dict):\n",
    "        # If data is a single dictionary, return the 'oxygen tolerance' value\n",
    "        return data['oxygen tolerance']\n",
    "    elif isinstance(data, list):\n",
    "        # If data is a list of dictionaries, extract 'oxygen tolerance' from each dict\n",
    "        return [d['oxygen tolerance'] for d in data]\n",
    "    else:\n",
    "        # Handle other unexpected data types\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f438f29f-8a42-40a9-8540-8c8d0fb883c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['aerobe' 'anaerobe' 'microaerophile' 'obligate aerobe'\n",
      " 'facultative anaerobe' 'obligate anaerobe' 'facultative aerobe'\n",
      " 'aerotolerant' 'microaerotolerant']\n"
     ]
    }
   ],
   "source": [
    "o2_tolerance_series = o2_tolerance_filtered_df['oxygen tolerance']\n",
    "\n",
    "oxygen_tolerance_values = o2_tolerance_series.apply(extract_oxygen_tolerance).explode()\n",
    "\n",
    "# Get unique values using unique() method\n",
    "unique_oxygen_tolerance_values = oxygen_tolerance_values.dropna().unique()\n",
    "\n",
    "print(unique_oxygen_tolerance_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e9931a9-b27c-4054-825a-669798bcd58e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
